src/dag_s.f90

```diff
@@ -106,55 +106,30 @@ end function toposort
             end do
           end associate
         end do
-            
+
         is_sorted_and_acyclic = .true.
       end block
-    
+
     end associate
-    
+
   end procedure
 
   module procedure construct_from_json
     type(fallible_json_value_t) :: maybe_vertices
 
+    maybe_vertices = json_object%get_element("vertices")
     associate(errors => maybe_vertices%errors())
       call assert(.not. errors%has_any(), "dag_s construct_from_json: .not. errors%has_any()", char(errors%to_string()))
     end associate
 
     select type (vertices => maybe_vertices%value_())
     type is (json_array_t)
-
-      associate(nvertices => vertices%length())
-        allocate(dag%vertices(nvertices))
-        block
-          integer i
-          call dag%vertices%set_vertex_id( [(i,i=1,nvertices)] )
-        end block
-      end associate
-
-      block
-        integer i
-        do i = 1, vertices%length()
-          associate(maybe_vertex => vertices%get_element(i))
-            associate(errors => maybe_vertex%errors())
-              call assert(.not. errors%has_any(), "dag_s construct_from_json: .not. errors%has_any()", char(errors%to_string()))
-            end associate
-
-            select type (vertex_json => maybe_vertex%value_())
-            class default
-              call assert(.false., "dag_s construct_from_json: vertex was not an object", char(vertex_json%to_compact_string()))
-            type is (json_object_t)
-              associate(dag_vertex => vertex_t(vertex_json))
-                call dag%vertices(i)%set_edges(dag_vertex%edges)
-              end associate
-            end select
-          end associate
-        end do
-      end block
+      dag%vertices = vertex_t(vertices%get_elements())
```

Comment:
@everythingfunctional this is truly elegant and is exactly the sort of thing I had hoped was possible.